python/smqtk/algorithms/relevancy_index/logistic_reg.py

```diff
@@ -0,0 +1,246 @@
+import collections
+import os.path as osp
+
+import numpy
+import six
+from six.moves import range, zip
+from six.moves import cPickle as pickle
+
+from smqtk.algorithms.relevancy_index import RelevancyIndex
+from smqtk.utils.parallel import parallel_map
+
+try:
+    import sklearn
+    from sklearn.linear_model import LogisticRegression
+except ImportError:
+    LogisticRegression = None
+    sklearn = None
+
+
+class LogisticRegRelevancyIndex (RelevancyIndex):
+    """
+    Uses Logistic regression python interface, using cosine distance,
+    to implement IQR ranking.
+    """
+
+    # Dictionary of parameter/value pairs that will be passed to libSVM during
+    # the model trail phase. Parameters that are flags, i.e. have no values,
+    # should be given an empty string ('') value.
+    LR_TRAIN_PARAMS = {
+        "penalty": "l2",
+        "dual": True,  
+        "class_weight": "balanced",
+        "random_state": 2,
+        "multi_class": "ovr",
+        "warm_start": False,
+        "n_jobs": 1,
+    }
+
+    @classmethod
+    def is_usable(cls):
+        """
+        Check whether this implementation is available for use.
+        Required valid presence of svm and svmutil modules
+        :return:
+            Boolean determination of whether this implementation is usable.
+        :rtype: bool
+        """
+        return LogisticRegression and sklearn
+
+    def __init__(self, descr_cache_filepath=None, autoneg_select_ratio=1,
+                 multiprocess_fetch=False, cores=None):
+        """
+        Initialize a new or existing index.
+        TODO ::
+        - input optional known background descriptors, i.e. descriptors for
+            things that would otherwise always be considered a negative example.
+        :param descr_cache_filepath: Optional path to store/load descriptors
+            we index.
+        :type descr_cache_filepath: None | str
+        :param autoneg_select_ratio: Number of maximally distant descriptors to
+            select from our descriptor cache for each positive example provided
+            when no negative examples are provided for ranking.
+            This must be an integer. Default value of 1. It is advisable not to
+            make this value too large.
+        :type autoneg_select_ratio: int
+        :param multiprocess_fetch: Use multiprocessing vs threading when
+            fetching descriptor vectors during ``build_index``. Default is
+            False.
+        :type multiprocess_fetch: bool
+        :param cores: Cores to use when performing parallel operations. A value
+            of None means to use all available cores.
+        :type cores: int | None
+        """
+        super(LogisticRegRelevancyIndex, self).__init__()
+
+        self.descr_cache_fp = descr_cache_filepath
```

Comment:
Nothing ever writes to this path here. Given the actual usage of relevancy indexes its probably OK if this not take a possible cache since its never used that way; they're always ephemeral.

Reply:
Removed cache.